piece-retriever/bin/piece-retriever.js

```diff
@@ -71,43 +71,65 @@ export default {
       // Timestamp to measure file retrieval performance (from cache and from SP)
       const fetchStartedAt = performance.now()
 
-      const [{ serviceProviderId, serviceUrl, dataSetId }, isBadBit] =
-        await Promise.all([
-          getStorageProviderAndValidatePayer(
-            env,
-            payerWalletAddress,
-            pieceCid,
-            env.ENFORCE_EGRESS_QUOTA,
-          ),
-          env.BAD_BITS_KV.get(`bad-bits:${await getBadBitsEntry(pieceCid)}`, {
-            type: 'json',
-          }),
-        ])
+      const [retrievalCandidates, isBadBit] = await Promise.all([
+        getRetrievalCandidatesAndValidatePayer(
+          env,
+          payerWalletAddress,
+          pieceCid,
+          env.ENFORCE_EGRESS_QUOTA,
+        ),
+        env.BAD_BITS_KV.get(`bad-bits:${await getBadBitsEntry(pieceCid)}`, {
+          type: 'json',
+        }),
+      ])
 
       httpAssert(
         !isBadBit,
         404,
         'The requested CID was flagged by the Bad Bits Denylist at https://badbits.dwebops.pub',
       )
 
-      httpAssert(
-        serviceProviderId,
-        404,
-        `Unsupported Service Provider: ${serviceProviderId}`,
-      )
+      if (retrievalCandidates.length === 0) {
+        const response = new Response('No retrieval candidate found', {
+          status: 404,
+        })
+        setContentSecurityPolicy(response)
+        return response
+      }
 
+      let retrievalCandidate
       let retrievalResult
+      const retrievalAttempts = []
 
-      try {
-        retrievalResult = await retrieveFile(
-          ctx,
-          serviceUrl,
-          pieceCid,
-          request,
-          env.ORIGIN_CACHE_TTL,
-          { signal: request.signal },
+      while (retrievalCandidates.length > 0) {
+        const retrievalCandidateIndex = Math.floor(
+          Math.random() * retrievalCandidates.length,
         )
-      } catch {}
+        retrievalCandidate = retrievalCandidates[retrievalCandidateIndex]
+        retrievalAttempts.push(retrievalCandidate)
+        retrievalCandidates.splice(retrievalCandidateIndex, 1)
+        console.log('Attempting retrieval', retrievalCandidate)
+        try {
+          retrievalResult = await retrieveFile(
+            ctx,
+            retrievalCandidate.serviceUrl,
+            pieceCid,
+            request,
+            env.ORIGIN_CACHE_TTL,
+            { signal: request.signal },
+          )
+          if (retrievalResult.response.ok) {
+            break
+          }
+        } catch {
+          console.log('Retrieval attempt failed', {
+            retrievalCandidate,
+            willRetry: retrievalCandidates.length > 0,
+          })
+        }
```

Comment:
We need to log a message also when we get a retrievalResult, but the response was not `ok`.

To make troubleshooting easier, let's include the caught error message or the response status code in the log please.


```suggestion
          if (retrievalResult.response.ok) {
            break
          }
          console.log(`Retrieval attempt failed: HTTP ${retrievalResult.response.status}', {
            retrievalCandidate,
            willRetry: retrievalCandidates.length > 0,
          })
        }
        } catch {
          console.log('Retrieval attempt failed: ${err.message ?? err}', {
            retrievalCandidate,
            willRetry: retrievalCandidates.length > 0,
          })
        }
```